src/_pytest/terminal.py

```diff
@@ -803,31 +806,36 @@ def getreports(self, name):
         return values
 
     def summary_warnings(self):
-        if self.hasopt("w"):
-            all_warnings = self.stats.get(
-                "warnings"
-            )  # type: Optional[List[WarningReport]]
-            if not all_warnings:
-                return
+        if not self.hasopt("w"):
+            return
 
-            final = hasattr(self, "_already_displayed_warnings")
-            if final:
-                warning_reports = all_warnings[self._already_displayed_warnings :]
-            else:
-                warning_reports = all_warnings
-            self._already_displayed_warnings = len(warning_reports)
-            if not warning_reports:
-                return
+        all_warnings = self.stats.get("warnings")  # type: Optional[List[WarningReport]]
```

Comment:
It looks like starting with `self.stats = defaultdict(list)` would allow us to simplify a lot of other code... the similar invocations all have `self.stats.get(..., [])` at the very least.  Happy to leave that to your judgement though.